File: sdk/rust/src/artifact/bun.rs
use crate::{
    api::artifact::ArtifactSystem::{
        Aarch64Darwin, Aarch64Linux, UnknownSystem, X8664Darwin, X8664Linux,
    },
    artifact::{step, system, Artifact, ArtifactSource},
    context::ConfigContext,
};
use anyhow::{bail, Result};
use indoc::formatdoc;

/// Canonical pin for the build-target `bun` runtime.
pub const DEFAULT_BUN_VERSION: &str = "1.3.10";

/// Build-target `bun` JavaScript runtime, fetched as a prebuilt binary release.
pub struct Bun {
    version: String,
}

impl Default for Bun {
    fn default() -> Self {
        Self {
            version: DEFAULT_BUN_VERSION.to_string(),
        }
    }
}

impl Bun {
    /// Creates a builder pinned to [`DEFAULT_BUN_VERSION`].
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Overrides the `bun` release version to fetch.
    #[must_use]
    pub fn with_version(mut self, version: &str) -> Self {
        self.version = version.to_string();
        self
    }

    /// Builds the `bun` artifact for the context's target system.
    ///
    /// # Errors
    ///
    /// Returns an error if the target system has no known `bun` release, or if registering
    /// the source or artifact with the build context fails.
    pub async fn build(self, context: &mut ConfigContext) -> Result<String> {
        let name = "bun";

        let system = context.get_system();

        let source_target = match system {
            Aarch64Darwin => "darwin-aarch64",
            Aarch64Linux => "linux-aarch64",
            X8664Darwin => "darwin-x64",
            X8664Linux => "linux-x64-baseline",
            UnknownSystem => bail!("unsupported {name} system: {}", system.as_str_name()),
        };

        let source_version = &self.version;
        let source_path =
            format!("https://sdk.vorpal.build/source/bun-{source_version}-{source_target}.zip");

        let source = ArtifactSource::new(name, source_path.as_str()).build();

        let step_script = formatdoc! {"
            mkdir -p \"$VORPAL_OUTPUT/bin\"
            cp -p \"./source/{name}/bun-{source_target}/bun\" \"$VORPAL_OUTPUT/bin/bun\"
            chmod +x \"$VORPAL_OUTPUT/bin/bun\"
        "};
        let steps = vec![step::shell(context, &[], &[], step_script, &[]).await?];

        Artifact::new(name, steps, system::SYSTEMS)
            .with_aliases(vec![format!("{name}:{source_version}")])
            .with_sources(vec![source])
            .build(context)
            .await
    }
}

File: sdk/typescript/src/artifact/bun.ts
import type { ConfigContext } from "../context.js";
import { Artifact, ArtifactSource } from "../artifact.js";
import { shell } from "./step.js";
import { getSystemStr, SYSTEMS } from "../system.js";

export const DEFAULT_BUN_VERSION = "1.3.10";

/**
 * Builder for the Bun runtime artifact.
 *
 * Mirrors Rust `Bun` struct in `sdk/rust/src/artifact/bun.rs`.
 * Downloads and extracts the Bun binary from a zip archive.
 */
export class Bun {
  private version: string;

  constructor() {
    this.version = DEFAULT_BUN_VERSION;
  }

  withVersion(version: string): this {
    this.version = version;
    return this;
  }

  async build(context: ConfigContext): Promise<string> {
    const name = "bun";
    const system = getSystemStr(context.getSystem());

    let sourceTarget: string;

    switch (system) {
      case "aarch64-darwin":
        sourceTarget = "darwin-aarch64";
        break;
      case "aarch64-linux":
        sourceTarget = "linux-aarch64";
        break;
      case "x86_64-darwin":
        sourceTarget = "darwin-x64";
        break;
      case "x86_64-linux":
        sourceTarget = "linux-x64-baseline";
        break;
      default:
        throw new Error(`unsupported ${name} system: ${system}`);
    }

    const sourceVersion = this.version;
    const sourcePath = `https://sdk.vorpal.build/source/bun-${sourceVersion}-${sourceTarget}.zip`;

    const source = new ArtifactSource(name, sourcePath).build();

    const stepScript = `mkdir -p "$VORPAL_OUTPUT/bin"
cp -p "./source/${name}/bun-${sourceTarget}/bun" "$VORPAL_OUTPUT/bin/bun"
chmod +x "$VORPAL_OUTPUT/bin/bun"
`;
    const steps = [await shell(context, [], [], stepScript, [])];

    return new Artifact(name, steps, SYSTEMS)
      .withAliases([`${name}:${sourceVersion}`])
      .withSources([source])
      .build(context);
  }
}
